### ISSUE TITLE
RFC: combo feature options

### ISSUE THREAD

**Issue by nirbheek** [mesonbuild/meson#4566]
It's a really common use-case to have something like:

```meson
option('3d-backend', type : 'combo', choices : ['auto', 'gl', 'gles', 'vulkan'], value : 'auto')
```
However, this is pretty annoying to implement in the build files themselves because you can't use the same constructs as feature options ala `required: get_option('3d-backend')`.

Would be nice to be able to do something like:

```meson
option('3d-backend', type : 'combo', choices : ['auto', 'gl', 'gles', 'vulkan'], value : 'auto')
...
gl_dep = dependency('gl', required : get_option('3d-backend:gl')
vulkan_dep = dependency('vulkan', required : get_option('3d-backend:vulkan')
```

This syntax should be ok because `:` is not allowed in option names. We can reuse the combo option type because we don't really need anything extra in the option definition.

**Comment by nirbheek** [comment]
This, however, is a bit incomplete ATM because it doesn't handle the case when building a `3d-backend` is itself optional. So we probably need a new option type that also has `disabled` as a valid value.

**Comment by dcbaker** [comment]
I'd *love* to have this for mesa, but I need the ability to key the auto value off of OS and CPU arch, that might be a separate PR, but it would be nice to at least keep it in mind.

**Comment by xclaesse** [comment]
> Would be nice to be able to do something like:
> 
> ```meson
> # `value:` can be any of the choices or `disabled` or `auto`
> option('3d-backend', type : 'feature-combo', choices : ['gl', 'gles', 'vulkan'], value : 'auto')
> ...
> gl_dep = dependency('gl', required : get_option('3d-backend:gl'))
> vulkan_dep = dependency('vulkan', required : get_option('3d-backend:vulkan'))
> ```

I'm not sure to follow that that example is supposed to do. Let's assume both gl and vulkan are available, and I build with `-D3d_backend=gl` then vulkan_dep would be not-found? If 3d-backend is auto then both gl_dep and vulkan_dep are found? 

One of the big reason we have `feature` is for the default_features global switch used by distro to not have surprises, how would you suggest integrating with that?

But I totally +1 the idea of combo-features, I just want to make sure we think about all cases ;D

**Comment by dcbaker** [comment]
While we're at it, a `list-features` would also be useful :)

**Comment by xclaesse** [comment]
Another question: Can I set `-D3d_backends=disabled` and none of them will be found?  I guess `-D3d_backends=enabled` wouldn't make sense, but then what if I set `-D3d_backends=auto -Dauto_features=enabled` ?

**Comment by xclaesse** [comment]
Well, I guess `-Dauto_features=enabled` could mean all of them must be found. But in practice I'm not sure that's convenient :/

**Comment by xclaesse** [comment]
Yeah, the more I think about it the more I like it, with my definition above.

`feature-array` would be defined in the same way, `get_option('3d-backend:gl')` would be `enabled` if `gl` is in the array, `disabled` otherwise. So you could do `-D3d_backends=disabled` or `-D3d_backends=gl,gles`, etc.

**Comment by nirbheek** [comment]
> but I need the ability to key the auto value off of OS and CPU arch, that might be a separate PR, but it would be nice to at least keep it in mind

Yes, it would be great to be able to define this in a declarative fashion. I keep coming across this use-case in GStreamer and GNOME, but I haven't thought of a good syntax for it yet.

> Another question: Can I set `-D3d_backends=disabled` and none of them will be found?

`feature-combo` would only accept one value, and `enabled` would not be a valid value since it's not one of the values in the choices. `disabled` would be a valid value to disable the option entirely. If the option is mandatory and cannot be disabled, you should use `combo`.

`-D3d_backends=auto -Dauto_features=enabled` should behave the same as it does now with `feature` options. I don't see how it's any different.

For options where multiple backends can be enabled at the same time (`feature-array`), I agree that we can probably use the syntax you've mentioned, and also extend it a bit to allow `get_option('3d-backend:gl,vulkan')` to check multiple options at the same time.

**Comment by xclaesse** [comment]
> -D3d_backends=auto -Dauto_features=enabled should behave the same as it does now with feature options. I don't see how it's any different.

Well, you said `enabled` is not a valid value for a feature-combo, so what do we do when a feature-combo is set to `auto` and auto_features is set to `enabled` ? I think it makes sense for feature-combo to have implicitly the extra 3 choices auto/enabled/disabled.

**Comment by dcbaker** [comment]
I think it makes sense for the feature-combo to return a `Feature` object internally, so that it can be queried with the feature methods, ala:

```meson
a = get_option('3d-backend:gl')
if a.enabled()
  ...
elif a.disabled()
  ...
endif
```

But I agree that a feature-combo should have the same semantics as a regular combo, only one option should be selectable. If you want more than one option that feels like it should use the `feature-list` I proposed.

**Comment by xclaesse** [comment]
+1 for `get_option('3d-backend:gl')` to return a `FeatureOptionHolder`, that's what I had in mind too.

I agree having `3d-backend=enabled` mean that all choices are enabled is unlikely to be what you want. But we still need to define what happens in that case. Ideally it should mean that we'll try all choices and at least one of them must work, but that's impossible to implement that way in a generic way, I think.

**Comment by nirbheek** [comment]
`3d-backend=enabled` should just be an error. `-D3d_backends=auto -Dauto_features=enabled` would also be an error in that case. You are required to provide a valid non-auto value to `3d_backends`.

**Comment by xclaesse** [comment]
Right, that's probably the easiest solution here. +1 for that behaviour then.

**Comment by emersion** [comment]
FWIW this is a pattern we use in wlroots: https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/013f121f455efcc1e68e98c60c74aadfb51fd98e/meson_options.txt#L5-8

It's pretty complicated and error-prone to manually handle all of the cases: https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/013f121f455efcc1e68e98c60c74aadfb51fd98e/render/allocator/meson.build

It would help a lot to have built-in support for this in Meson.

**Comment by xclaesse** [comment]
> It's pretty complicated and error-prone to manually handle all of the cases: https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/013f121f455efcc1e68e98c60c74aadfb51fd98e/render/allocator/meson.build

I think the part that's not covered with current proposal is there:

```
if 'auto' in allocators and get_option('auto_features').enabled()
	allocators = ['gbm']
```

That won't be allowed with current proposal:

> `3d-backend=enabled` should just be an error. `-D3d_backends=auto -Dauto_features=enabled` would also be an error in that case. You are required to provide a valid non-auto value to `3d_backends`.

Maybe we could have a way to provide the default value when the option is set to `enabled`, like: `option('allocators', type: 'feature-array', enabled: 'gbm')` ?

Nitpicking: I'm not super fan of the syntax `get_option('3d-backend:gl')` because the `:` separator is already used for `subproject:option` in other places, so it could be confusing to have another meaning here? Maybe have a dedicated kwarg, or even method like `get_option('3d-backend', for: 'gl')` or `get_option('3d-backend').for('gl')`?

Naming is hard as always.

**Comment by dcbaker** [comment]
I completely forgot that we'd talked about this before, I opened https://github.com/mesonbuild/meson/issues/10624 to talk about this with a slightly different syntax and way that you work with it.
